### PR TITLE
feat: implement limited support for Unix sockets

### PR DESCRIPTION
**Pull request by doudou**

### CODE REVIEW

**Review by jhonasiv — APPROVED** [review]

File: src/Driver.cpp
Comment (wvmcastro) [inline src/Driver.cpp:1]: I did not expect udp would be so heavy against plain sockets

File: src/IOStream.cpp [hunk 1]
```diff
@@ -277,3 +279,207 @@ size_t UDPServerStream::write(uint8_t const* buffer, size_t buffer_size)
...
+    if (m_client_stream) {
+        return m_client_stream->waitRead(timeout);
+    }
+
+    base::Time now = base::Time::now();
+    base::Time deadline = now + timeout;
+    while (now <= deadline) {
+        if (!m_fd_stream.waitRead(deadline - now)) {
+            return false;
+        }
```
Comment (wvmcastro) [inline src/IOStream.cpp]: did you mean
```suggestion
    base::Time deadline = base::Time::now() + timeout;
    while (base::Time::now() <= deadline) {
        if (!m_fd_stream.waitRead(timeout)) {
            return false;
        }
```
or it should be a plain `while(true)` ?

File: src/IOStream.hpp [hunk 2]
```diff
@@ -104,6 +107,61 @@ namespace iodrivers_base
...
+        size_t write(uint8_t const* buffer, size_t buffer_size) override;
+
+    protected:
+        /** Internal implementation of recvfrom to allow for mocking */
+        virtual std::pair<ssize_t, int> recvfrom(uint8_t* buffer,
+            size_t buffer_size,
+            int flags,
+            sockaddr* s_other,
+            socklen_t* s_len);
+        /** Internal implementation of recvfrom to allow for mocking */
```
Comment (wvmcastro) [inline src/IOStream.hpp]:
```suggestion
        /** Internal implementation of sendto to allow for mocking */
```

File: src/IOStream.cpp [hunk 3]
```diff
@@ -277,3 +279,207 @@ size_t UDPServerStream::write(uint8_t const* buffer, size_t buffer_size)
...
+pair<ssize_t, int> UnixDatagramStream::sendto(uint8_t const* buffer, size_t buffer_size)
+{
+    ssize_t ret = ::sendto(m_fd, buffer, buffer_size, 0, reinterpret_cast<sockaddr const*>(&m_si_other), m_si_other_len);
+    return make_pair(ret, errno);
+}
+
+size_t UnixDatagramStream::write(uint8_t const* buffer, size_t buffer_size)
+{
+    if (!m_has_other)
+        return buffer_size;
```
Comment (wvmcastro) [inline src/IOStream.cpp:454]: forgot the curly braces; shouldn't this return be zero?

File: src/IOStream.cpp (hunk 1 above)
Comment (doudou) [reply src/IOStream.cpp]: Thanks ! The while loop is unneeded.

File: src/IOStream.cpp (hunk 3 above)
Comment (doudou) [reply src/IOStream.cpp:454]: The behaviour is to ignore the data (same than with UDPServer). Returning zero would make iodrivers_base generate a timeout.

The implementation of UnixServerStream was actually wrong in respect to this behaviour. Fixed in the latest push.

Comment (doudou) [reply src/IOStream.cpp:454]: About the curly braces ... this is old code. I wanted to get this PR through before I reformat the whole file.

**Review by wvmcastro — APPROVED** [review]

**Pull request merged**